gadget/install/partition.go

```diff
@@ -268,9 +285,37 @@ func ensureNodesExistImpl(dss []gadget.OnDiskStructure, timeout time.Duration) e
 	return nil
 }
 
-// reloadPartitionTable instructs the kernel to re-read the partition
+// reloadPartitionTableWithDeviceRescan instructs the kernel to re-read the
+// partition table of a given block device via a workaround proposed for a
+// specific device in the form of executing the equivalent of:
+// bash -c "echo 1 > /sys/block/sd?/device/rescan"
+func reloadPartitionTableWithDeviceRescan(device string) error {
+	disk, err := disks.DiskFromDeviceName(device)
+	if err != nil {
+		return err
+	}
+
+	rescanFile := filepath.Join(disk.KernelDevicePath(), "device", "rescan")
+
+	logger.Noticef("reload partition table via rescan file %s for device %s as indicated by gadget", rescanFile, device)
+	f, err := os.OpenFile(rescanFile, os.O_WRONLY, 0644)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	// this could potentially fail with strange sysfs errno's since rescan isn't
+	// a real file
```

Comment:
is this operation async or synchronous? do we need to check for something to know that the rescan has happened?

Reply:
sd driver seems to be doing it synchronously https://elixir.bootlin.com/linux/latest/source/drivers/scsi/sd.c#L3275 looking at some other code calling scsi_rescan_device() (which is also called from the sysfs handler), it seems to be expected that this call is synchronous. However, the scanning in itself can be async, AFAICT it's handled by module parameter. I tried to dig into block layer, it's a bit convoluted, but the events to reread the partition table seem to be handled in a workqueue, so I would expect that that even through a SCSI level rescan complets, the block layer may take a while to add partitions.

Reply:
FWIW partprobe source code may provide more hints as to how userspace should behave

Reply:
well to be clear, even if the operation is asynchronous this call is immediately followed by two checks afterwards:
* we call `udevadm settle --timeout=180`
* then we wait for up to 5 seconds for the device nodes in /dev/sda3 etc to show up

Do we think these checks are insufficient? We do these checks regardless of whether we call partx or device/rescan like this. I believe the main reason we have those checks is because even when we use partx it shows up asynchronously so unless we think this device/rescan thing is much slower than using partx normally I think we are okay.

FWIW I could not find anywhere in partprobe (or GNU parted codebase) where `device/rescan` is used so I'm not sure what to follow there. In that code they are doing ioctl's with BKPLG there.

Reply:
I see